lib/type/number.ex

```diff
@@ -11,9 +11,115 @@ defmodule Litmus.Type.Number do
   ]
 
   @type t :: %__MODULE__{
-          min: non_neg_integer,
-          max: non_neg_integer,
+          min: non_neg_integer | nil,
+          max: non_neg_integer | nil,
           integer: boolean,
           required: boolean
         }
+
+  alias Litmus.Required
+
+  @spec validate_field(t, binary, map) :: {:ok, map} | {:error, binary}
+  def validate_field(type, field, data) do
+    with {:ok, data} <- Required.validate(type, field, data),
+         {:ok, data} <- convert(type, field, data),
+         {:ok, data} <- min_validate(type, field, data),
+         {:ok, data} <- max_validate(type, field, data),
+         {:ok, data} <- integer_validate(type, field, data) do
+      {:ok, data}
+    else
+      {:error, msg} -> {:error, msg}
+    end
+  end
+
+  @spec string_to_number(binary) :: number | :error
+  defp string_to_number(str) do
+    modified_str =
+      try do
+        String.to_integer(str)
+      rescue
+        ArgumentError ->
+          try do
+            String.to_float("0" <> str)
+          rescue
+            ArgumentError ->
+              :error
+          end
+      end
+
+    modified_str
+  end
+
+  @spec convert(t, binary, map) :: {:ok, map} | {:error, binary}
+  defp convert(%__MODULE__{}, field, params) do
+    cond do
+      !Map.has_key?(params, field) ->
+        {:ok, params}
+
+      is_number(params[field]) ->
+        {:ok, params}
+
+      is_binary(params[field]) ->
+        modified_value = string_to_number(params[field])
+
+        case modified_value do
+          :error ->
+            {:error, "#{field} must be a number"}
+
+          _ ->
+            {:ok, Map.put(params, field, modified_value)}
+        end
```

Comment:
Let's see if we can rewrite this to pull the `case` statement out of the `cond` statement. We generally want to avoid nested conditionals if possible.